fix(AIP-149): differentiate field presence and behavior (#1143)
googleapis/aip: aip/general/0149.md

value and unset most of the time; if an alternative design does not require
such a distinction, it is usually preferred. In practice, this means `optional`
**should** only ever be used for integers and floats.

**Important:** Tracking field presence is *not* the same as documenting API
field behavior as defined in [AIP-203][]. For example, a field labeled with

> `optional` for presence tracking **may** also be annotated as

Reviewer: Should there be a cross-reference to http://cloud/apis/design/design_patterns.md#optional_primitive_fields ?

Author: bah! Shouldn't *that* link to *this* AIP? Anyways, discussion of primitives, etc. is part of the guidance above, I don't think we need to discuss it further in this note. Is that fair?

Reviewer 2: Yeah, I'd rather move more to the AIPs. it's more actively maintained.

The bifurcation cloud APIs and AIPs is a known issue though - when we have more breathing room we should probably tackle that.

`google.api.field_behavior = REQUIRED` if the field must be set. If you only

> want to document the server perceived behavior of a field, read AIP-203.

Reviewer 3: Link to AIP-203?

Author: I linked to it earlier in the paragraph and I'm not sure if I should get link happy. I will include it for good measure but curious what you think is good practice?

Author: added the link

Reviewer 3: Don't mind either way - I hadn't spotted the fact that you'd linked to it already in the same paragraph, sorry.

Author: All good! Probably better to link anyways.

## Rationale

> ### `optional` and field behavior

Reviewer: Flip these. Or flip their references in the paragraph below.

IOW, this heading talks about `optional` first, but in the paragraph below, it refers optional as the latter.

Author: Good call. Done.

The field behavior annotation and `optional` label are not mutually exclusive,
because they address different problems. The former,
`google.api.field_behavior`, focuses on communicating the server's perception of
a field within the API e.g. if it is required or not, if it is immutable, etc.
The latter, proto3's `optional`, is a wire format and code generation option
that is strictly for toggling field presence tracking. While it might be
confusing for a field to be simultaneously annotated with
`google.api.field_behavior = REQUIRED` and labeled as `optional`, they are

> unrelated in practice and are reasonable to be used together.

Reviewer 3: ```suggestion
unrelated in practice and can reasonably be used together.
```

Reviewer 3: (Or "and it is reasonable for them to be used together".)
